### DISCUSSION TITLE
Beads in PraxisLIVE?

### DISCUSSION THREAD

**Discussion by antonlapov** [praxis-live/support#130] [Ideas]
Hi. I came across PraxisLIVE after having some experience of audio programing using Beads library (in fact I learnt about existence of this project from Ollie Bown in beads google group forum). I am curious is it a relevant idea, in general, to code using Beads as a third-party library inside (audio?) patch/root custom component ? Is it possible to pass along audio (and control) signal generated by Beads from components outputs? Thanks!

**Comment by neilcsmith-net** [comment]
Well, anything is possible.  Not necessarily easy!  PraxisLIVE uses a UGen library called Pipes that is somewhat similar to Beads - https://github.com/jaudiolibs/pipes  It can also be used as a standalone library.  One option is possibly to wrap various Bead UGens in a Pipe UGen.  The FFT component in the examples projects (eg. live coder) uses a custom Pipe class, that is actually calling FFT code ported from Beads.  It might be a starting point.

What is missing from Pipes that you might want to use from Beads?

**Reply by antonlapov** [reply]
Thank you for your quick reply. I will look into mentioned example and try to wrap my head around this.

It's hard to say in terms of comparison of these two libraries. I've just already known a bit how beads work, I am familiar with syntax and conventions, enjoy using wide range of pre-made UGens (e.g. working with Envelopes generators (especially, let's say, "beading" events by triggering different functions at the the end of enevelopes) etc.. There are two books and some useful examples accumulated since the beginning of the project..So I think it might be an good idea to use that legacy in live-coding environment. 
I developed just several simple A/V projects inside Processing IDE with beads (and on another IDE) but I guess that with rising of complexity of CG code the overall performance would decrease. That's why I am quite interested in your forest-of-actors concept. But anyway I am guessing: I am very high-level user and don't possess any hardcore  coding skills

On the other hand, having much more experience with live/node-based environment (vvvv,Pd) I like an idea of JIT compilation and on the fly prototyping etc..anyways..
Thanks for your work! 
I hope to dive into Praxis more when I have more time.

**Reply by neilcsmith-net** [reply]
Well, PraxisLIVE tends to use a slightly different approach for things like envelopes.  There are `Property` objects that support key frame animation and function calls at the end which can be used for controlling UGen parameters.  They're not audio specific, so can be used in video and other graph types too.

You can see the same concept in the read-me example on the Pipes repo -

```
@Inject Property sweep, env;
...
sweep.animator().whenDone(p -> p.to(8000, 65).in(4, 0.2).easeInOut());
sweep.link(filter::frequency);
env.link(d -> osc.gain(d * d * d * d));
```

When used inside PraxisLIVE, injected properties retain their animation as you live code, while also disconnecting and reconnecting them to the ugens automatically.  One issue with Beads could be handling listeners / callbacks across code changes - it's possible but a bit more convoluted.

In my own live-coding I often tend to work more low level than this, though, at the per-sample level :smile: - https://youtu.be/tkJxd2IOQQY

Let me know how you get on.  I'm just back to working a lot on PraxisLIVE again after a bit of a hiatus.  v6 should be released in the next week or so.

**Reply by antonlapov** [reply]
Thanks for the reply!
I'll keep my eye on your future release and try to play around with Pipes, Property objects.
